src/bvals/comms/combined_buffers.cpp

```diff
@@ -0,0 +1,414 @@
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <cstdio>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "bvals/comms/bvals_utils.hpp"
+#include "bvals/comms/combined_buffers.hpp"
+#include "bvals/neighbor_block.hpp"
+#include "coordinates/coordinates.hpp"
+#include "interface/variable.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "utils/communication_buffer.hpp"
+
+namespace parthenon {
+
+CombinedBuffersRank::CombinedBuffersRank(int o_rank, BoundaryType b_type, bool send)
+    : other_rank(o_rank), b_type(b_type), sender(send), buffers_built(false) {
+
+  int tag = 1234 + static_cast<int>(GetAssociatedSender(b_type));
+  if (sender) {
+    message = com_buf_t(tag, Globals::my_rank, other_rank, comm_,
+                        [](int size) { return std::vector<int>(size); });
+  } else {
+    message = com_buf_t(
+        tag, other_rank, Globals::my_rank, comm_,
+        [](int size) { return std::vector<int>(size); }, true);
+  }
+  PARTHENON_REQUIRE(other_rank != Globals::my_rank, "Should only build for other ranks.");
+}
+
+void CombinedBuffersRank::AddSendBuffer(int partition, MeshBlock *pmb,
+                                        const NeighborBlock &nb,
+                                        const std::shared_ptr<Variable<Real>> &var,
+                                        BoundaryType b_type) {
+  if (current_size.count(partition) == 0) current_size[partition] = 0;
+  auto &cur_size = current_size[partition];
+  combined_info[partition].push_back(
+      BndId::GetSend(pmb, nb, var, b_type, partition, cur_size));
+  cur_size += combined_info[partition].back().size();
+}
+
+bool CombinedBuffersRank::TryReceiveBufInfo(Mesh *pmesh) {
+  PARTHENON_REQUIRE(!sender, "Trying to receive on a combined sender.");
+  if (buffers_built) return buffers_built;
+
+  bool received = message.TryReceive();
+  if (received) {
+    auto &mess_buf = message.buffer();
+    int npartitions = mess_buf[0];
+    // Unpack into per combined buffer information
+    int idx{nglobal};
+
+    for (int p = 0; p < npartitions; ++p) {
+      const int partition = mess_buf[idx++];
+      const int nbuf = mess_buf[idx++];
+      const int total_size = mess_buf[idx++];
+      combined_buffers[partition] =
+          CommBuffer<buf_t>(913 + partition, other_rank, Globals::my_rank, comm_);
```

Comment:
What in the world is `913`?

Reply:
I was just worried there would be tag collisions and 913 was the first number that my mental RNG produced.

Reply:
now have moved away from this, using independent communicators.